core/src/main/scala/org/apache/spark/ui/memory/MemoryTab.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui.memory
+
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.executor.{TransportMetrics, ExecutorMetrics}
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster.ExecutorInfo
+import org.apache.spark.ui.{SparkUITab, SparkUI}
+
+private[ui] class MemoryTab(parent: SparkUI) extends SparkUITab(parent, "memory") {
+  val memoryListener = parent.memoryListener
+  val progressListener = parent.jobProgressListener
+  attachPage(new MemoryPage(this))
+  attachPage(new StageMemoryPage(this))
+}
+
+/**
+ * :: DeveloperApi ::
+ * A SparkListener that prepares information to be displayed on the MemoryTab
+ */
+@DeveloperApi
+class MemoryListener extends SparkListener {
+  type ExecutorId = String
+  val activeExecutorIdToMem = new HashMap[ExecutorId, MemoryUIInfo]
+  val removedExecutorIdToMem = new HashMap[ExecutorId, MemoryUIInfo]
+  // latestExecIdToExecMetrics include all executors that is active and removed.
+  // this may consume a lot of memory when executors are changing frequently, e.g. in dynamical
+  // allocation mode.
+  val latestExecIdToExecMetrics = new HashMap[ExecutorId, ExecutorMetrics]
+  // stagesIdToMem a map maintains all executors memory information of each stage,
+  // the Map type is [(stageId, attemptId), Seq[(executorId, MemoryUIInfo)]
+  val activeStagesToMem = new HashMap[(Int, Int), HashMap[ExecutorId, MemoryUIInfo]]
+  val completedStagesToMem = new HashMap[(Int, Int), HashMap[ExecutorId, MemoryUIInfo]]
+
+  override def onExecutorMetricsUpdate(event: SparkListenerExecutorMetricsUpdate): Unit = {
+    val executorId = event.execId
+    val executorMetrics = event.executorMetrics
+    val memoryInfo = activeExecutorIdToMem.getOrElseUpdate(executorId, new MemoryUIInfo)
+    memoryInfo.updateExecutorMetrics(executorMetrics)
+    activeStagesToMem.map {stageToMem =>
+      if (stageToMem._2.contains(executorId)) {
+        val memInfo = stageToMem._2.get(executorId).get
+        memInfo.updateExecutorMetrics(executorMetrics)
```

Comment:
you should use `activeStagesToMem.foreach` instead of `.map` since you're doing this for side-effects, not for the result of the `map`.  Also a little nicer if you use a `case` to extract the right part of the tuple:

```scala
activeStagesToMem.foreach { case (_, stageMetrics) =>
  if (stageMetrics.contains(executorId)) {
   ...
```